Dense no cython extra functions

MrRobot2211 opened the pull request:
This PR will contain the extra function of the `CuPyDense` class which are not part of QuTiP's `data.Data`

MrRobot2211 committed:
first

MrRobot2211 committed:
stubbed all functions

MrRobot2211 committed:
fixed alt init function:
now they start by calling __new__

MrRobot2211 committed:
clean-up

MrRobot2211 committed:
fixes constructors

MrRobot2211 committed:
implements  and tests imul,mul,matmul,truediv,itruediv

hodgestar commented:
@MrRobot2211 I'm going to hold off reviewing this until #4 is merged or closer to merging, but please feel free to continue here as you like. :D

MrRobot2211 committed:
ADDS BENCHMARKS

MrRobot2211 committed:
Merge branch 'main' of https://github.com/qutip/qutip-cupy into dense_no_cython_extra_functions

MrRobot2211 committed:
\corrects merge

MrRobot2211 committed:
fixes indentation on dense

MrRobot2211 committed:
adds div, mul, matmul and itruediv basic tests

MrRobot2211 committed:
removed spurious benchmark

hodgestar commented:
@MrRobot2211 One #19 is merged, could you merge main into this branch and fix up any things flake8 or black complain about and then ping me when you're ready for this to be reviewed?

MrRobot2211 commented:
> @MrRobot2211 Once #19 is merged, could you merge main into this branch and fix up any things flake8 or black complain about and then ping me when you're ready for this to be reviewed?

 
Yes, of course @hodgestar. There are many functions missing here, as well.  After I merge the changes and complete all the functions  I'll ping you.

MrRobot2211 committed:
Merge branch 'main' of https://github.com/qutip/qutip-cupy into dense_no_cython_extra_functions

MrRobot2211 committed:
reanmes internal specialized functions
to follow the convention <operation>_<class>

MrRobot2211 committed:
makes dense_function.py file

MrRobot2211 committed:
refactor specialized constructor.
they will now use the raw constructor.

MrRobot2211 committed:
add diags constructor

MrRobot2211 committed:
formats

MrRobot2211 committed:
formats

MrRobot2211 committed:
formats

MrRobot2211 committed:
dispatch imul,mul and matmul

MrRobot2211 committed:
formats

MrRobot2211 committed:
fix neg_cupydense name

MrRobot2211 committed:
dispatch matmul

MrRobot2211 committed:
dispatch add

MrRobot2211 committed:
dispatch sub

MrRobot2211 commented:
Well #15 is already handled.

MrRobot2211 committed:
dispatches specialized constructors

MrRobot2211 committed:
stubs qutip/reshape.py functions

MrRobot2211 committed:
add trace_cupydense

MrRobot2211 committed:
stub ptrace_cupydense

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+        return self
+
+    def __neg__(self):
+
```
```suggestion
```
Remove blank line.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(left._cp - right._cp)
+
+    # def __dealloc__(self):
+    #     if self._deallocate and self.data != NULL:
+    #         PyDataMem_FREE(self.data)
+
```
```suggestion
```
We can remove `__dealloc__` -- this class is pure Python.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+    #         PyDataMem_FREE(self.data)
+
+
+# @TOCHECK  emti and empty_like are not dispatched in qutip
+# if we do not use them internally here we should remove them
+def empty(rows, cols, fortran):
+    """
+    Return a new Dense type of the given shape, with the data allocated but
+    uninitialised.
+    """
+    order = "F" if fortran else "C"
+    cparr = cp.empty(shape=(rows, cols), dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
+
+
+def empty_like(other, fortran):
+    """
+    Return a new Dense type of the same shape as the given array.
+    """
+    order = "F" if fortran else "C"
+    cparr = cp.empty_like(other, dtype=cp.complex128, order=order)
+    return CuPyDense._raw_cupy_constructor(cparr)
```
I suggest we remove `empty` and `empty_like` from this PR, and then add create an issue in this repo to add them and to add the dispatchers to qutip dev.major.

MrRobot2211 replied:
When you say the dispatchers  you refer to  `empty`'s and `empty-like`'s dispatcher right? 

Shouldn't we move all the constructors to a separate PR then?

hodgestar replied:
Now we are both confused. :)

In this PR, you register specialisations for `diag`, `identity` and `zeros` -- so those are already used.

You don't register specialisations for `empty` and `emtpy_like` so those are unused (and hence I am suggesting not including them in the qutip-cupy code until they can be used).

MrRobot2211 replied:
Ah, I see, you are riight the thing is they are not dispathced in `dev.major` either,  see [here](https://github.com/qutip/qutip/blob/dev.major/qutip/core/data/constant.py). I think they are for internal use, maybe I should ask @jakelishman .

jakelishman replied:
Not every function is meant to have a dispatcher - `csr.sort` won't either, because it only makes sense for CSR.  `empty` and `empty_like` might not make sense for CuPy, but for `Dense` and `CSR` they're used heavily in internal Cython code to implement various features such as `copy` and as the allocation base for `add`, and so on.  I don't see any particular use for them in Python-only space, though, so there's no dispatcher.

hodgestar replied:
The use within `qutip-cupy` would be to create an empty matrix on the GPU without having to copy it from the CPU, but I think we can live without empty for the moment.

jakelishman replied:
It'll never make sense for TensorFlow or any potential symbolic data type, so I wouldn't think it'll ever become a dispatcher. I don't see a reason why you'd ever want the low-level object-creation routines to be abstractly dispatched.  How would you go about adding data in in the correct form if you don't know what type of object you've just created?  That's why the medium-level `diags`, `identity`, `one_element`, etc exist, and those guys do have dispatchers.

The expectation is that if you ever wanted to call the super low-level `csr` functions from Python space, it would be in something that looked a little like:
```python
out = csr.copy_structure(other)
out_scipy = out.as_scipy(full=True)
out_scipy.data[:] = ...
return out
```
where you're returning a concrete type, not an abstract one.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+    is passed, then the result will be `scale` times the identity.
+    """
+    order = "F" if fortran else "C"
+    if scale != 1:
```
```suggestion
    if scale == 1:
```
We should also add tests for `scale=1`, scale=`2`, scale=`1+5j`.

MrRobot2211 replied:
changed and the tests are being habdled by qutip

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+    return CuPyDense._raw_cupy_constructor(cparr)
+
+
+def Dense_from_csr(CSR_matrix, fortran=False):
+    pass
+
+
```
```suggestion
```
Added the suggestion to remove the empty function.

hodgestar commented on src/qutip_cupy/dense.py:
```diff
@@ -116,17 +321,82 @@ def cupydense_from_dense(dense):
 
 
-def cpd_adjoint(cpd_array):
+def adjoint_cupydense(cpd_array):  # noqa: E302
```
E302 is "Expected two blank lines". Why do we ignore that here? It looks like there are two blank lines above this function definition? Same question for other functions.

MrRobot2211 replied:
I might have had a mix-up here, but I thought that we inclined to having smaller spacing of short funtions, so that the files do not become to long and is eadier for reading.

jakelishman replied:
I would say either commit to the code style, or disable that warning globally.  Personally I don't see any problem with the extra spacing - it's better to be consistent, and have a tool you can use to say "this is always the right way to do it" than to say "this is mostly how we do it, but then there's a bunch of exceptions".  In the latter case, you just end up with a higgledy piggledy mess, and you spend a bunch of time arguing in code review with contributors that "this patch is formatted really badly, please fix", and they say "but _your_ code has all sorts of exceptions".  Better to just be hard-nosed about it and say, "it might not be universally perfect, but consistent and unalienable is better in the long run".

hodgestar reviewed (commented):
I know this is still in progress, but I did a partial review of what is here to give some earlier feedback and hopefully save some time later.

Broader comments:
* We need a lot more tests (probably a few for each new specialization function)
* The implementation of `diags` looks very odd in places, but I assume it's a work in progress.

MrRobot2211 commented on src/qutip_cupy/__init__.py:
```diff
@@ -36,10 +36,22 @@
+data.add.add_specialisations([(CuPyDense, CuPyDense, CuPyDense, cd.add_cupydense)])
+data.sub.add_specialisations([(CuPyDense, CuPyDense, CuPyDense, cd.sub_cupydense)])
+
+data.diag.add_specialisations([(CuPyDense, cd.diags)])
```
I made the same question on issue #15 see [this](f.append(experiment_dict) I am sticking to the naming used in qutip dev.major, but I think the proper thing would be to extend the naming convention as I suggest there,

MrRobot2211 replied:
In fact `diags` is the name given in `dev.major`  but I am happier with `diag`.

MrRobot2211 replied:
Although if we were to apply the same principle as on PR #15 , we should keep`diags` because a later developer might be confused by the name chages between `qutip`  and `qutip-cupy`

jakelishman replied:
It's `diags` rather than `diag` because that matched `qutip.qdiags`, and it can accept a list of diagonals rather than just one (as opposed to `np.diag`).  I was mostly just matching pre-existing stuff in QuTiP, but in general I think `diags` makes sense.  I commented on the historical reasons for the naming convention somewhere else, but the gist was that originally I assumed that anything defined in `csr.pyx` or `dense.pyx` would only ever be accessed by qualified access through that module, and having the additional type information was superfluous, consequently adding noise to the code when you're reading it.  It would be clear, say, that you needed a type suffix on a `norm.l2_*` function, though, because those don't have the type in the module name.

From the dispatcher user's perspective, it really doesn't matter what they're called, because they never see the raw objects anyway.  I'm not against renaming things like `csr.diags` to `diags_csr`, and having data's `__init__.pxd` and `__init__.py` import those select names from `CSR`.

MrRobot2211 commented:
> I know this is still in progress, but I did a partial review of what is here to give some earlier feedback and hopefully save some time later.
> 
> Broader comments:
> 
> * We need a lot more tests (probably a few for each new specialization function)
> * The implementation of `diags` looks very odd in places, but I assume it's a work in progress.

Thank you @hodgestar.
Yes, my next step was to bring tests from `dev.major`.

On the `diags` function there are some minimal changes to work with `cupy arrays` instead of `numpy arrays`, but most of the code is borrowedt from `dev.major`. In any case, as it is not being tested I wouldn't put my hands on the fire for this one.

MrRobot2211 committed:
Borrows tests from qutip

MrRobot2211 committed:
adds iadd

MrRobot2211 committed:
clean up

MrRobot2211 committed:
add shape checks to sub

MrRobot2211 committed:
qutip style tests: add mul,matmult,transpose

MrRobot2211 commented:
I am migrating all test to qutip style because they are thorough, and because it really makes the coding faster.
I am a little bit uncomfortable with some of the Exception raised checks though. 
For example there are checks for the shapes in `add` and `sub` but not in `matmul`

MrRobot2211 committed:
add transpose specialization

MrRobot2211 commented:
To make `transpose` pass QuTiP's tests I have to call `.item()` (see 
567b944) this is because `cp.trace` returns a [1,1] ` CuPyArray . this forces comunication between CPU and GPU and affects the timings substatially. This same issue will arise in the implementation of `expect` also, maybe we will have to do something similar to what Asier has done to allow `mul` by [1,1] arrays.

MrRobot2211 committed:
lint formatter conflict solved
delete  , in string concatenation

MrRobot2211 committed:
dispatches reshape

MrRobot2211 committed:
clean up tydi up

MrRobot2211 committed:
empty,zeros,identity constructors pass tests

MrRobot2211 committed:
clean-up

MrRobot2211 committed:
clean-up

MrRobot2211 committed:
tests: add random _cupydense function

MrRobot2211 commented:
Important note there is no test for the  reshape dispatch in QuTiP, so I will create aPR to add it so I can just inherit it here as I did with the rest of the mathematical functions

MrRobot2211 committed:
diags constructor pass test

MrRobot2211 committed:
clean-up

MrRobot2211 committed:
corrects capitalization in docstring

hodgestar commented on tests/test_dense.py:
```diff
@@ -50,3 +51,135 @@ def test_adjoint(self, shape):
+
+def test_true_div(shape):
+
+    from qutip.core import data
```
These imports should just be at the top of the test module.

MrRobot2211 committed:
remove emty and empty-like
constructors and tersts

MrRobot2211 committed:
tests: refactor make only one
import of qutip.data

MrRobot2211 committed:
add  todo comment in trace
change in calling .item

MrRobot2211 committed:
sort imports properly

MrRobot2211 committed:
add comment on single precision random generator

MrRobot2211 committed:
clean-up

MrRobot2211 committed:
dispatch tidyup

MrRobot2211 commented:
@hodgestar I think there is no harm on leaving the original dense_tests as they are checking the functions firectly on the class and not on the dispatched specializations.

MrRobot2211 committed:
filter qutip matplotlib warnings

hodgestar commented on src/qutip_cupy/dense_functions.py:
```diff
@@ -0,0 +1,29 @@
+
+
+def reshape_cupydense(cp_arr, n_rows_out, n_cols_out):
+
+    return CuPyDense(cp_arr, (n_rows_out, n_cols_out))
```
I think here `cp_arr` will be a `CuPyDense` already, so this function should be `CuPyDense._raw_cupy_constructor(cupy.reshape(cp_arr._cp, (n_rows_out, n_cols_out))`?

MrRobot2211 replied:
yes, good catch

hodgestar commented on tests/test_dense.py:
```diff
@@ -1,7 +1,17 @@
 import numpy as np
 import pytest
 
+with warnings.catch_warnings():
+    warnings.filterwarnings(
+        action="ignore", category=UserWarning, message=r"matplotlib not found:"
+    )
+
+    from qutip.core import data
+
+from qutip_cupy import CuPyDense  # noqa:E402
+from qutip_cupy import dense  # noqa:E402
```
```suggestion
from qutip_cupy import CuPyDense
from qutip_cupy import dense

# import qutip after qutip_cupy to supress the matplotlib import warning
from qutip.core import data
```
I think importing qutip after qutip_cupy will avoid the warning and save copying and pasting the warning filter in all the tests.

hodgestar commented on tests/test_dense.py:
```diff
@@ -50,3 +58,120 @@ def test_adjoint(self, shape):
+    array = np.random.uniform(size=shape) + 1.0j * np.random.uniform(size=shape)
+
+    cup_arr = CuPyDense(array)
+    cpdense_tr = cup_arr / 2.0
+    qtpdense_tr = data.Dense(array) / 2.0
+
+    np.testing.assert_array_equal(cpdense_tr.to_array(), qtpdense_tr.to_array())
```
```suggestion
    cpdense_over_2 = cup_arr / 2.0
    qtpdense_over_2 = data.Dense(array) / 2.0

    np.testing.assert_array_equal(cpdense_over_2.to_array(), qtpdense_over_2.to_array())
```
Fix copy and pasted names.

hodgestar commented on tests/test_dense.py:
```diff
@@ -50,3 +58,120 @@ def test_adjoint(self, shape):
+
+    array = np.random.uniform(size=shape) + 1.0j * np.random.uniform(size=shape)
+
+    cpdense_tr = CuPyDense(array).__itruediv__(2.0)
+    qtpdense_tr = data.Dense(array).__itruediv__(2.0)
+
+    np.testing.assert_array_equal(cpdense_tr.to_array(), qtpdense_tr.to_array())
```
Please fix copy and pasted variable names.

It's probably also a good idea to instead or also test
```python
cpdense = CuPyDense(array)
cpdense /= 2
```
to ensure that the inplace operator works as expected.

hodgestar commented on tests/test_dense.py:
```diff
@@ -50,3 +58,120 @@ def test_adjoint(self, shape):
+
+    array = np.random.uniform(size=shape) + 1.0j * np.random.uniform(size=shape)
+
+    cpdense_tr = CuPyDense(array).__mul__(2.0 + 1.0j)
+    qtpdense_tr = data.Dense(array).__mul__(2.0 + 1.0j)
+
+    np.testing.assert_array_equal(cpdense_tr.to_array(), qtpdense_tr.to_array())
```
Fix c&p variable names.

hodgestar commented on tests/test_dense.py:
```diff
@@ -50,3 +58,120 @@ def test_adjoint(self, shape):
+
+    array = np.random.uniform(size=shape) + 1.0j * np.random.uniform(size=shape)
+
+    cpdense_tr = CuPyDense(array).__mul__(2.0 + 1.0j)
+    qtpdense_tr = data.Dense(array).__mul__(2.0 + 1.0j)
+
+    np.testing.assert_array_equal(cpdense_tr.to_array(), qtpdense_tr.to_array())
```
Fix c&p variable names and the name of the test. Also test `CuPyDense(array) * (2.0 + 1.0j)`.

hodgestar commented on tests/test_mathematics.py:
```diff
@@ -0,0 +1,93 @@
+import cupy as cp
+import pytest
+
+with warnings.catch_warnings():
+    warnings.filterwarnings(
+        action="ignore", category=UserWarning, message=r"matplotlib not found:"
+    )
+    import qutip.tests.core.data.test_mathematics as test_tools
+
+from qutip_cupy import dense  # noqa:E402
+from qutip_cupy import dense_functions as cdf  # noqa:E402
+from qutip_cupy import CuPyDense  # noqa:E402
```
Same comment about moving the qutip import after importing qutip_cupy and removing the `# noqa`s and warning filter.

MrRobot2211 replied:
After checking we do not need this filter in particular, and there is no noqa comment needed then. 
But @hodgestar , from past comments , is the correct order base python libraries, external libraries, our own modules?

hodgestar replied:
By default, yes -- Python standard library, then third party packages, then this package. However, sometimes exceptions can be made if needed in strange cases such as this, which is why I suggested keeping a comment.

Note that if you run all the tests with `pytest tests` then only the first test file would need the strange ordering but I recommend keeping the strange ordering in all test files so that they can also be run individually with `pytest tests/test_foo.py`.

hodgestar reviewed (changes requested):
Reviewed!

I left quite a few comments, but all of them are requests for relatively small changes, so hopefully they can be implemented or discussed quickly.

MrRobot2211 committed:
fix dispatch zeros

MrRobot2211 committed:
switch qutip and qutip_cupy import order

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
Merge branch 'dense_no_cython_extra_functions' of https://github.com/MrRobot2211/qutip-cupy into dense_no_cython_extra_functions

MrRobot2211 commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +93,211 @@ def adjoint(self):
+
+
+def Dense_from_csr(CSR_matrix, fortran=False):
+    pass
```
yes and already done

MrRobot2211 committed:
removed unused warnings

MrRobot2211 committed:
correc typo on zeros

MrRobot2211 committed:
fix variable names in tests

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
fix itrue_test_div names

MrRobot2211 committed:
fix test mul names

MrRobot2211 committed:
test_mathmatics remove warning
filter anf noqa comment

MrRobot2211 commented on src/qutip_cupy/dense.py:
```diff
@@ -116,17 +292,120 @@ def cupydense_from_dense(dense):
+    -------
+    out : CUPyDense
+    """
+    _check_same_shape(left, right)
```
The check is required because the tests in qutip look for a specific exception. If you remove it all incorrect_shape tests fail

MrRobot2211 replied:
Besides I think that if you do not check it the default behaviour in numpy and cupy is to broadcast the smallest onto the largest.

hodgestar replied:
Re broadcasting: Good point. Let's keep the checks!

MrRobot2211 committed:
diags refactor diagonal_length variable

MrRobot2211 committed:
fix dense_function module docstring

MrRobot2211 committed:
fix reshape dispatch function

MrRobot2211 committed:
fix diags formatting

MrRobot2211 commented on src/qutip_cupy/dense_functions.py:
```diff
@@ -0,0 +1,29 @@
+# Contains functions for dense_cupy, this are the same functions that
+# are defined ouside the dense.pyx file
```
Fixed and turned into docstring

MrRobot2211 committed:
changed import sorting order
this avoids breaking test run  on this single file
when warnings in qutip are found

MrRobot2211 commented on src/qutip_cupy/__init__.py:
```diff
@@ -23,6 +23,7 @@
 # qutip_cupy imports need to be after the cupy import check above
 from .version import version as __version__  # noqa: E402
 from . import dense as cd  # noqa: E402
+from . import dense_functions as cdf  # noqa: F401
```
They should all be E402. You are right.

MrRobot2211 committed:
fix noqa comment in __init__

MrRobot2211 committed:
simplify imul function

Co-authored-by: Simon Cross <hodgestar+github@gmail.com>

MrRobot2211 committed:
fix dispatch matmul to reassign to out

MrRobot2211 committed:
Merge branch 'dense_no_cython_extra_functions' of https://github.com/MrRobot2211/qutip-cupy into dense_no_cython_extra_functions

MrRobot2211 commented on src/qutip_cupy/dense.py:
```diff
@@ -116,17 +292,120 @@ def cupydense_from_dense(dense):
+    if out:
+        return (left @ right) * scale + out
+    else:
+        return (left @ right) * scale
```
Good catch.  I reordered the if though.

MrRobot2211 committed:
fix dense matmul test

MrRobot2211 commented on src/qutip_cupy/dense.py:
```diff
@@ -93,6 +95,180 @@ def adjoint(self):
+    def __add__(left, right):
+        if not isinstance(left, CuPyDense) or not isinstance(right, CuPyDense):
+            return NotImplemented
+        return CuPyDense._raw_cupy_constructor(cp.add(left._cp, right._cp))
```
No good reason. I am changing it.

MrRobot2211 committed:
refactor __add__  to have the same
style as matmul

MrRobot2211 committed:
remove blank line

MrRobot2211 commented:
@hodgestar  I think I took on all the changes, except this one https://github.com/qutip/qutip-cupy/pull/10#discussion_r677497874 for which I still need need your feedback.

Some of your suggestions I could not apply directly because the caused breaking changes. Tell me if my changes are to your liking.  

After this one there are 4 very shorts PRs ready.

hodgestar reviewed (approved):
@MrRobot2211 Thank you for making all the changes! Looks great.

Merge when you're ready.
